File: bindings/python/src/kortecx/wait.py
"""``wait`` orchestration — turn an async run handle into a single result.

This composes existing RPCs client-side, exactly like the ``kx`` CLI
(``wait.rs``): poll ``GetProjection`` until the target Mote is terminal, then
``GetContent`` its committed result. Two strategies share one :class:`WaitOutcome`:

* **poll** (default, CLI-parity) — re-read the projection every 250 ms.
* **events** (opt-in, lower latency) — subscribe to ``StreamEvents`` and react to
  the terminal Mote's delta as it lands (sub-second), resuming on a CatchupRequired
  drop. This is strictly more efficient than the poll and is forward-compatible
  with the live tail.

Both sync (``poll_*`` / ``events_*``) and async (``apoll_*`` / ``aevents_*``)
variants exist; they operate on the same generated ``KxGatewayStub`` (the channel
kind — sync or aio — decides whether a call returns a value or an awaitable).
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

import grpc

from . import types
from .errors import from_rpc_error
from .v1 import gateway_pb2 as _g

#: Polling cadence — matches the CLI's bounded backoff (never a busy spin).
POLL_INTERVAL = 0.25


class WaitState(str, Enum):
    """The terminal disposition of a waited-on run."""

    COMMITTED = "COMMITTED"
    FAILED = "FAILED"
    RUNNING = "RUNNING"  # timed out, still in progress, resumable


@dataclass
class WaitOutcome:
    """Server-derived ids + the terminal disposition (+ result on commit)."""

    instance_id: bytes
    terminal_mote_id: bytes
    state: WaitState
    result_ref: Optional[bytes] = None
    payload: Optional[bytes] = None


def _terminal(instance_id: bytes, mote_id: bytes, state: WaitState) -> WaitOutcome:
    return WaitOutcome(instance_id=instance_id, terminal_mote_id=mote_id, state=state)


def _snapshot_result_ref(m: "_g.MoteSnapshot") -> Optional[bytes]:
    return m.result_ref if m.HasField("result_ref") else None


# --- sync (poll) --------------------------------------------------------------


def _get_projection(stub, md, instance_id: bytes) -> "_g.ProjectionView":
    try:
        return stub.GetProjection(_g.GetProjectionRequest(instance_id=instance_id), metadata=md)
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


def _get_content(stub, md, instance_id: bytes, content_ref: bytes) -> bytes:
    try:
        return stub.GetContent(
            _g.GetContentRequest(content_ref=content_ref, instance_id=instance_id),
            metadata=md,
        ).payload
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


def _committed_outcome(stub, md, instance_id, mote_id, result_ref) -> WaitOutcome:
    payload = _get_content(stub, md, instance_id, result_ref) if result_ref else None
    return WaitOutcome(
        instance_id=instance_id,
        terminal_mote_id=mote_id,
        state=WaitState.COMMITTED,
        result_ref=result_ref,
        payload=payload,
    )


def poll_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    """Poll until ``terminal_mote_id`` is terminal (the ``invoke`` path)."""
    deadline = time.monotonic() + timeout
    while True:
        view = _get_projection(stub, md, instance_id)
        m = next((x for x in view.motes if x.mote_id == terminal_mote_id), None)
        if m is not None:
            if types.is_committed(m.state):
                return _committed_outcome(
                    stub, md, instance_id, terminal_mote_id, _snapshot_result_ref(m)
                )
            if not types.is_pending(m.state):
                return _terminal(instance_id, terminal_mote_id, WaitState.FAILED)
        if time.monotonic() >= deadline:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        time.sleep(POLL_INTERVAL)


#: The branches a ReAct turn settles to (vs the live "pending"/"tool" states).
_REACT_ANSWER = "answer"
_REACT_DEAD = "dead_lettered"


def _list_react_turns(stub, md, instance_id: bytes):
    try:
        resp = stub.ListReactTurns(
            _g.ListReactTurnsRequest(instance_id=instance_id), metadata=md
        )
        return list(resp.turns)
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


def _projection_result_ref(stub, md, instance_id: bytes, mote_id: bytes) -> Optional[bytes]:
    view = _get_projection(stub, md, instance_id)
    m = next((x for x in view.motes if x.mote_id == mote_id), None)
    return _snapshot_result_ref(m) if m is not None else None


def poll_react_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    """Wait for a ReAct CHAIN to settle (the ``invoke`` react path).

    A react chain has no statically-known terminal Mote: the run-salted turn-0 id
    the gateway hands back never matches the committed turn id, and the settled
    Answer turn isn't known until the model emits it. So completion is observed via
    ``ListReactTurns`` — the chain is done when a turn settles to ``answer`` (the
    final answer; resolve its committed content) or ``dead_lettered`` (terminal
    failure). This is the durable, server-derived signal (the runtime's own
    "resume with get_projection / events" hint, made the default for react).
    """
    deadline = time.monotonic() + timeout
    while True:
        turns = _list_react_turns(stub, md, instance_id)
        answer = next((t for t in turns if t.branch == _REACT_ANSWER), None)
        if answer is not None:
            rr = _projection_result_ref(stub, md, instance_id, answer.turn_mote_id)
            return _committed_outcome(stub, md, instance_id, answer.turn_mote_id, rr)
        dead = next((t for t in turns if t.branch == _REACT_DEAD), None)
        if dead is not None:
            return _terminal(instance_id, dead.turn_mote_id, WaitState.FAILED)
        if time.monotonic() >= deadline:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        time.sleep(POLL_INTERVAL)


def poll_any(stub, md, instance_id, timeout) -> WaitOutcome:
    """Poll until ANY Mote commits (the ``submit`` path — no terminal id)."""
    deadline = time.monotonic() + timeout
    while True:
        view = _get_projection(stub, md, instance_id)
        committed = next((x for x in view.motes if types.is_committed(x.state)), None)
        if committed is not None:
            return _committed_outcome(
                stub, md, instance_id, committed.mote_id, _snapshot_result_ref(committed)
            )
        if view.motes and all(not types.is_pending(x.state) for x in view.motes):
            first = view.motes[0].mote_id
            return _terminal(instance_id, first, WaitState.FAILED)
        if time.monotonic() >= deadline:
            return _terminal(instance_id, b"", WaitState.RUNNING)
        time.sleep(POLL_INTERVAL)


# --- sync (events) ------------------------------------------------------------


def events_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    """Wait via the live event stream (lower latency than the poll).

    Subscribes from ``seq 0`` (so an already-committed terminal Mote is seen in the
    catch-up replay), watches for the terminal Mote's committed/failed delta, and
    resumes from the last cursor on a CatchupRequired drop.
    """
    deadline = time.monotonic() + timeout
    cursor = 0
    while True:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        req = _g.StreamEventsRequest(instance_id=instance_id, since_seq=cursor)
        try:
            for frame in stub.StreamEvents(req, metadata=md, timeout=remaining):
                for d in frame.deltas:
                    which = d.WhichOneof("kind")
                    if which == "committed" and d.committed.mote_id == terminal_mote_id:
                        rr = d.committed.result_ref or None
                        return _committed_outcome(stub, md, instance_id, terminal_mote_id, rr)
                    if which == "failed" and d.failed.mote_id == terminal_mote_id:
                        return _terminal(instance_id, terminal_mote_id, WaitState.FAILED)
                cursor = frame.next_seq
        except grpc.RpcError as e:
            code = e.code()
            if code == grpc.StatusCode.RESOURCE_EXHAUSTED:
                continue  # CatchupRequired: resume from the last cursor
            if code == grpc.StatusCode.DEADLINE_EXCEEDED:
                return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
            raise from_rpc_error(e) from e


# --- async (poll) -------------------------------------------------------------


async def _aget_projection(stub, md, instance_id: bytes) -> "_g.ProjectionView":
    try:
        return await stub.GetProjection(
            _g.GetProjectionRequest(instance_id=instance_id), metadata=md
        )
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


async def _aget_content(stub, md, instance_id, content_ref) -> bytes:
    try:
        resp = await stub.GetContent(
            _g.GetContentRequest(content_ref=content_ref, instance_id=instance_id),
            metadata=md,
        )
        return resp.payload
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


async def _acommitted_outcome(stub, md, instance_id, mote_id, result_ref) -> WaitOutcome:
    payload = await _aget_content(stub, md, instance_id, result_ref) if result_ref else None
    return WaitOutcome(
        instance_id=instance_id,
        terminal_mote_id=mote_id,
        state=WaitState.COMMITTED,
        result_ref=result_ref,
        payload=payload,
    )


async def apoll_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    loop = asyncio.get_event_loop()
    deadline = loop.time() + timeout
    while True:
        view = await _aget_projection(stub, md, instance_id)
        m = next((x for x in view.motes if x.mote_id == terminal_mote_id), None)
        if m is not None:
            if types.is_committed(m.state):
                return await _acommitted_outcome(
                    stub, md, instance_id, terminal_mote_id, _snapshot_result_ref(m)
                )
            if not types.is_pending(m.state):
                return _terminal(instance_id, terminal_mote_id, WaitState.FAILED)
        if loop.time() >= deadline:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        await asyncio.sleep(POLL_INTERVAL)


async def _alist_react_turns(stub, md, instance_id: bytes):
    try:
        resp = await stub.ListReactTurns(
            _g.ListReactTurnsRequest(instance_id=instance_id), metadata=md
        )
        return list(resp.turns)
    except grpc.RpcError as e:
        raise from_rpc_error(e) from e


async def _aprojection_result_ref(stub, md, instance_id: bytes, mote_id: bytes) -> Optional[bytes]:
    view = await _aget_projection(stub, md, instance_id)
    m = next((x for x in view.motes if x.mote_id == mote_id), None)
    return _snapshot_result_ref(m) if m is not None else None


async def apoll_react_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    """Async mirror of :func:`poll_react_result` (the react ``invoke`` path)."""
    loop = asyncio.get_event_loop()
    deadline = loop.time() + timeout
    while True:
        turns = await _alist_react_turns(stub, md, instance_id)
        answer = next((t for t in turns if t.branch == _REACT_ANSWER), None)
        if answer is not None:
            rr = await _aprojection_result_ref(stub, md, instance_id, answer.turn_mote_id)
            return await _acommitted_outcome(stub, md, instance_id, answer.turn_mote_id, rr)
        dead = next((t for t in turns if t.branch == _REACT_DEAD), None)
        if dead is not None:
            return _terminal(instance_id, dead.turn_mote_id, WaitState.FAILED)
        if loop.time() >= deadline:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        await asyncio.sleep(POLL_INTERVAL)


async def apoll_any(stub, md, instance_id, timeout) -> WaitOutcome:
    loop = asyncio.get_event_loop()
    deadline = loop.time() + timeout
    while True:
        view = await _aget_projection(stub, md, instance_id)
        committed = next((x for x in view.motes if types.is_committed(x.state)), None)
        if committed is not None:
            return await _acommitted_outcome(
                stub, md, instance_id, committed.mote_id, _snapshot_result_ref(committed)
            )
        if view.motes and all(not types.is_pending(x.state) for x in view.motes):
            return _terminal(instance_id, view.motes[0].mote_id, WaitState.FAILED)
        if loop.time() >= deadline:
            return _terminal(instance_id, b"", WaitState.RUNNING)
        await asyncio.sleep(POLL_INTERVAL)


# --- async (events) -----------------------------------------------------------


async def aevents_result(stub, md, instance_id, terminal_mote_id, timeout) -> WaitOutcome:
    loop = asyncio.get_event_loop()
    deadline = loop.time() + timeout
    cursor = 0
    while True:
        remaining = deadline - loop.time()
        if remaining <= 0:
            return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
        req = _g.StreamEventsRequest(instance_id=instance_id, since_seq=cursor)
        call = stub.StreamEvents(req, metadata=md, timeout=remaining)
        try:
            async for frame in call:
                for d in frame.deltas:
                    which = d.WhichOneof("kind")
                    if which == "committed" and d.committed.mote_id == terminal_mote_id:
                        rr = d.committed.result_ref or None
                        return await _acommitted_outcome(
                            stub, md, instance_id, terminal_mote_id, rr
                        )
                    if which == "failed" and d.failed.mote_id == terminal_mote_id:
                        return _terminal(instance_id, terminal_mote_id, WaitState.FAILED)
                cursor = frame.next_seq
        except grpc.RpcError as e:
            code = e.code()
            if code == grpc.StatusCode.RESOURCE_EXHAUSTED:
                continue
            if code == grpc.StatusCode.DEADLINE_EXCEEDED:
                return _terminal(instance_id, terminal_mote_id, WaitState.RUNNING)
            raise from_rpc_error(e) from e


def collect_frames(frames: List["_g.EventFrame"]) -> List[types.Frame]:
    """Map raw proto frames to views (used by the events helpers + tests)."""
    return [types.Frame.from_proto(f) for f in frames]

File: bindings/typescript/src/wait.ts
/**
 * `wait` orchestration — turn a started run into a single result.
 *
 * Composes existing RPCs client-side, exactly like the `kx` CLI (`wait.rs`) and
 * the Python SDK: poll `GetProjection` until the target Mote is terminal, then
 * `GetContent` its committed result. Two strategies share one {@link WaitOutcome}:
 *
 * - **poll** (default, CLI-parity) — re-read the projection every 250 ms.
 * - **events** (opt-in, lower latency) — subscribe to `StreamEvents` and react to
 *   the terminal Mote's delta as it lands, resuming on a CatchupRequired drop.
 */

import { Code, ConnectError } from "@connectrpc/connect";
import type { Client } from "@connectrpc/connect";
import { fromRpcError, rpc } from "./errors.js";
import type { KxGateway } from "./gen/kortecx/v1/gateway_pb.js";
import { MoteSnapshotState } from "./gen/kortecx/v1/gateway_pb.js";

/** The terminal disposition of a waited-on run. */
export const WaitState = {
  Committed: "COMMITTED",
  Failed: "FAILED",
  /** timed out, still in progress, resumable. */
  Running: "RUNNING",
} as const;
export type WaitState = (typeof WaitState)[keyof typeof WaitState];

/** The two wait strategies. */
export type WaitMode = "poll" | "events";

/** Server-derived ids + the terminal disposition (+ result on commit). */
export interface WaitOutcome {
  instanceId: Uint8Array;
  terminalMoteId: Uint8Array;
  state: WaitState;
  resultRef?: Uint8Array;
  payload?: Uint8Array;
}

type Gateway = Client<typeof KxGateway>;

/** Polling cadence — matches the CLI's bounded backoff (never a busy spin). */
const POLL_INTERVAL_MS = 250;

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

function eq(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function isCommitted(s: MoteSnapshotState): boolean {
  return s === MoteSnapshotState.COMMITTED;
}

function isPending(s: MoteSnapshotState): boolean {
  return s === MoteSnapshotState.PENDING || s === MoteSnapshotState.SCHEDULED;
}

async function fetchContent(
  gw: Gateway,
  instance: Uint8Array,
  ref: Uint8Array,
): Promise<Uint8Array> {
  const blob = await rpc(gw.getContent({ contentRef: ref, instanceId: instance }));
  return blob.payload;
}

async function committedOutcome(
  gw: Gateway,
  instance: Uint8Array,
  mote: Uint8Array,
  ref: Uint8Array | undefined,
): Promise<WaitOutcome> {
  const hasRef = ref !== undefined && ref.length > 0;
  const payload = hasRef ? await fetchContent(gw, instance, ref) : undefined;
  return {
    instanceId: instance,
    terminalMoteId: mote,
    state: WaitState.Committed,
    resultRef: hasRef ? ref : undefined,
    payload,
  };
}

/** Poll until `terminal` is terminal (the `invoke` path). */
export async function pollResult(
  gw: Gateway,
  instance: Uint8Array,
  terminal: Uint8Array,
  timeoutMs: number,
): Promise<WaitOutcome> {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const view = await rpc(gw.getProjection({ instanceId: instance }));
    const m = view.motes.find((x) => eq(x.moteId, terminal));
    if (m !== undefined) {
      if (isCommitted(m.state)) {
        return committedOutcome(gw, instance, terminal, m.resultRef);
      }
      if (!isPending(m.state)) {
        return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Failed };
      }
    }
    if (Date.now() >= deadline) {
      return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Running };
    }
    await sleep(POLL_INTERVAL_MS);
  }
}

/** The branches a ReAct turn settles to (vs the live "pending"/"tool" states). */
const REACT_ANSWER = "answer";
const REACT_DEAD = "dead_lettered";

/**
 * gemma3 connector-tool-fire: under the Ollama non-strict UNION `format` a settled react
 * answer turn commits `{"answer": "…"}` instead of free prose; unwrap it to the plain text
 * a caller expects. Mirrors the Rust `kx_toolcall::extract_answer`, the `kx` CLI, and the
 * Python SDK — a byte-identical NO-OP for prose / llama.cpp answers or any body that is not
 * EXACTLY a single-key `{"answer": <string>}` object (presentation only).
 */
function extractAnswer(payload: Uint8Array | undefined): Uint8Array | undefined {
  if (payload === undefined) return payload;
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(payload).trim();
  } catch {
    return payload;
  }
  if (!text.startsWith("{")) return payload;
  let obj: unknown;
  try {
    obj = JSON.parse(text);
  } catch {
    return payload;
  }
  if (
    typeof obj === "object" &&
    obj !== null &&
    !Array.isArray(obj) &&
    Object.keys(obj).length === 1 &&
    typeof (obj as Record<string, unknown>).answer === "string"
  ) {
    return new TextEncoder().encode((obj as Record<string, string>).answer);
  }
  return payload;
}

/**
 * Wait for a ReAct CHAIN to settle (the `invoke` react path).
 *
 * A react chain has no statically-known terminal Mote: the run-salted turn-0 id
 * the gateway hands back never matches the committed turn id, and the settled
 * Answer turn isn't known until the model emits it. So completion is observed via
 * `ListReactTurns` — done when a turn settles to `answer` (resolve its committed
 * content) or `dead_lettered` (terminal failure). Mirrors the runtime's own
 * "resume with get_projection / events" hint (campaign finding F13).
 */
export async function pollReactResult(
  gw: Gateway,
  instance: Uint8Array,
  terminal: Uint8Array,
  timeoutMs: number,
  chainSalt: Uint8Array = new Uint8Array(0),
): Promise<WaitOutcome> {
  // PR-R1: scope the settle poll to THIS invocation's chain (serve shares one
  // journal/instance_id across every Invoke). An empty salt = instance-only scoping.
  const stepSalt = chainSalt.length > 0 ? chainSalt : undefined;
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const resp = await rpc(gw.listReactTurns({ instanceId: instance, stepSalt }));
    const answer = resp.turns.find((t) => t.branch === REACT_ANSWER);
    if (answer !== undefined) {
      const view = await rpc(gw.getProjection({ instanceId: instance }));
      const m = view.motes.find((x) => eq(x.moteId, answer.turnMoteId));
      const outcome = await committedOutcome(gw, instance, answer.turnMoteId, m?.resultRef);
      return { ...outcome, payload: extractAnswer(outcome.payload) };
    }
    const dead = resp.turns.find((t) => t.branch === REACT_DEAD);
    if (dead !== undefined) {
      return { instanceId: instance, terminalMoteId: dead.turnMoteId, state: WaitState.Failed };
    }
    if (Date.now() >= deadline) {
      return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Running };
    }
    await sleep(POLL_INTERVAL_MS);
  }
}

/** Poll until ANY Mote commits (the `submit` path — no terminal id). */
export async function pollAny(
  gw: Gateway,
  instance: Uint8Array,
  timeoutMs: number,
): Promise<WaitOutcome> {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const view = await rpc(gw.getProjection({ instanceId: instance }));
    const committed = view.motes.find((x) => isCommitted(x.state));
    if (committed !== undefined) {
      return committedOutcome(gw, instance, committed.moteId, committed.resultRef);
    }
    const first = view.motes[0];
    if (first !== undefined && view.motes.every((x) => !isPending(x.state))) {
      return { instanceId: instance, terminalMoteId: first.moteId, state: WaitState.Failed };
    }
    if (Date.now() >= deadline) {
      return { instanceId: instance, terminalMoteId: new Uint8Array(0), state: WaitState.Running };
    }
    await sleep(POLL_INTERVAL_MS);
  }
}

/** Wait via the live event stream (lower latency than the poll). */
export async function eventsResult(
  gw: Gateway,
  instance: Uint8Array,
  terminal: Uint8Array,
  timeoutMs: number,
): Promise<WaitOutcome> {
  const deadline = Date.now() + timeoutMs;
  let cursor = 0n;
  for (;;) {
    const remaining = deadline - Date.now();
    if (remaining <= 0) {
      return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Running };
    }
    try {
      for await (const frame of gw.streamEvents(
        { instanceId: instance, sinceSeq: cursor },
        { timeoutMs: remaining },
      )) {
        for (const d of frame.deltas) {
          if (d.kind.case === "committed" && eq(d.kind.value.moteId, terminal)) {
            const rr = d.kind.value.resultRef;
            return committedOutcome(gw, instance, terminal, rr.length > 0 ? rr : undefined);
          }
          if (d.kind.case === "failed" && eq(d.kind.value.moteId, terminal)) {
            return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Failed };
          }
        }
        cursor = frame.nextSeq;
      }
    } catch (e) {
      const ce = ConnectError.from(e);
      if (ce.code === Code.ResourceExhausted) continue; // CatchupRequired: resume from cursor
      if (ce.code === Code.DeadlineExceeded || ce.code === Code.Canceled) {
        return { instanceId: instance, terminalMoteId: terminal, state: WaitState.Running };
      }
      throw fromRpcError(e);
    }
    // The snapshot stream ended before the terminal committed — re-subscribe from
    // the cursor after a short backoff (avoids a hot loop; the terminal commits soon).
    await sleep(POLL_INTERVAL_MS);
  }
}
